### ISSUE TITLE
TFX support for TFDS

### ISSUE THREAD

**Issue by vijayphoenix** [tensorflow/datasets#2082]
**Is your feature request related to a problem? Please describe.**
[TFX](https://www.tensorflow.org/tfx) is an end-to-end platform for deploying production ML pipelines.
They have an [ExampleGen](https://www.tensorflow.org/tfx/guide/examplegen) to load data from various sources (databases, tfrecord,...):
It would be good to have ExampleGen support TFDS natively. 
This would unlock TFDS datasets for production models.

**Describe the solution you'd like**
A possible solution could be to reuse the existing TFRecord writer in TFDS and create an API 
like:
```python
example_gen = CreateExampleGenfromTFDS(name="cifar10")
```
which would somewhat mimic the following code:

**Sample Code using TFX with TFDS**

```python
import tensorflow_datasets as tfds
from tfx.utils.dsl_utils import external_input
from tfx.components.example_gen.import_example_gen.component import ImportExampleGen

# Generate TFRecords with TFDS
builder = tfds.builder("cifar10")
builder.download_and_prepare()

# Use the TFRecords to create a ExampleGen
examples = external_input(builder.data_dir)
example_gen = ImportExampleGen(input=examples)
```

**Describe alternatives you've considered**
* Reuse ExempleGen or develop our own

**Comment by Conchylicultor** [comment]
Thank you for looking into this Vijay.
I think there might be two parts we might want to provide for the users:
* `ExampleGen`: Read the data from TF-Record
* `TFT`: Decode the images, eventually cast values,... (our current `info.features.decode_examples`)

A first good step would be to write a full end-to-end TFX with a trainer to see how TFDS can be plugged into a dummy model in practice.

Once done, we can see which boilerplate parts can be simplified with utils functions.

One thing which seem missing from the code snippet you copied is that it's not clear how to specify the train/test splits.

Once the colab is done, we could commit the colab inside `docs/tfx.ipynb` to add a new `TFX` section to our website (similarly to our [overview](https://www.tensorflow.org/datasets/overview) colab).

You can submit a new colab here. Once done, similarly to the `docs/overview`.

**Comment by 8bitmp3** [comment]
Thanks @Conchylicultor. It would be awesome if `ExampleGen` and TFDS were easily interoperable. Also, more examples in small samples would help across the TFDS site, similar to what @vijayphoenix suggested. And, since there are so many moving parts in TFX, maybe you can use references in the future Colab (`docs/tfx.ipynb`) to/from various TFX/TFDS pages - the search engine would pick up on them, I guess (and hope we don't have to go outside of TF.org to GitHub and search stuff under Issues or PRs often to find solutions.) 

(Less related potential improvement suggestion: since the official TFDS guide on how to use params with MNIST is very general, it would be awesome to have more detailed docs similar to [`torchvision.datasets`](https://pytorch.org/docs/stable/torchvision/datasets.html). It would reduce the time spent loading, augmenting, wrangling etc if we know what we're doing from the get-go. cc @dynamicwebpaige)

**Comment by Conchylicultor** [comment]
> it would be awesome to have more detailed docs similar to torchvision.datasets. It would reduce the time spent loading, augmenting, wrangling etc if we know what we're doing from the get-go

Could you provide more context ? What do you mean by `params` ? Which types of doc would you like which isn't provided by https://www.tensorflow.org/datasets/overview, or https://www.tensorflow.org/datasets/keras_example ?

**Comment by carlthome** [comment]
>A first good step would be to write a full end-to-end TFX with a trainer to see how TFDS can be plugged into a dummy model in practice.

Has someone here done this first work now and can showcase a hello world perhaps?

Speaking as a very happy TFDS user for exploratory ML work, as I have a model I'm happy with I'm now looking for how programmatically scheduled workflows in TFX would play with TFDS artifacts in an easy and _intended_ way, particularly for non-static data that's ingested from a data warehouse (daily BigQuery partitions, joined with GCS blobs, in my case).

I realize that TFDS was designed for static datasets, iterated upon by humans during a model development and R&D stage, and it's great for that, so the above might be deemed to be feature creep. But it would be a killer feature if one could also take an existing dataset builder class and schedule it with some incremental versioning happening automagically per run, and let the current semver tag be like the "pipeline version" in TFX but with multiple runs associated with it.

Asking before I (re?)invent some custom TFX component for this.

**Comment by lxkain** [comment]
Just want to +1 that interoperability between tfx and tfds would be amazing.

**Comment by carlthome** [comment]
AFAIK, @OscarUtterback got initial boilerplate setup for consuming the TFDS TFRecord naming scheme in an TFX ExampleGen. Perhaps you could share it here? :wave:

**Comment by OscarUtterback** [comment]
Yep, this is done via Input.Split, using file-name patterns. Given that TFDS has generated a data set named `epidemic/tiny:1.0.0`, this will ingest the tfrecord files according to the split defined by the pattern parameter.

It's not as nice as using `tfds.load`, but it works.

```
from tfx.components.example_gen.import_example_gen.component import ImportExampleGen
from tfx.proto import example_gen_pb2
input_config = example_gen_pb2.Input(
    splits = [
        example_gen_pb2.Input.Split(name='train', pattern='epidemic-train*'),
        example_gen_pb2.Input.Split(name='test', pattern='epidemic-test*')
    ]
)
data_dir = "/home/oscar/data/epidemic/tiny/1.0.0"
example_gen = ImportExampleGen(input_base=data_dir, input_config=input_config) 
```

**Comment by lxkain** [comment]
Thank you for providing this example! It looks like this can take care of versioning, splits, and configuration by simply pointing at the right directory. However, as you hint at, the tfds.load argument also offers functionality in terms of acquiring just parts of the data (specified in absolute or percentage terms) - would this best be implemented using methods on the example_gen object?

**Comment by lxkain** [comment]
Cross-referencing:
https://github.com/tensorflow/tfx/issues/2027
